Adds --web flag to open codespace in vscode web

jackcmeyer opened the pull request:
Fixes #4985

- Adds `--web` flag to open the codespace in vscode web. 
- updates `codespaces/api/api.go` `Codespace` type to include `WebUrl`

jackcmeyer committed:
Adds --web flag to open codespace in vscode web

jackcmeyer commented on internal/codespaces/api/api.go:
```diff
@@ -183,6 +183,7 @@ type Codespace struct {
 	VSCSTarget                     string              `json:"vscs_target"`
 	PendingOperation               bool                `json:"pending_operation"`
 	PendingOperationDisabledReason string              `json:"pending_operation_disabled_reason"`
+	WebUrl                         string              `json:"web_url"`
```
Is adding `WebUrl` here the right path forward? Using this as the link to open in the browser.

luanzeba commented on pkg/cmd/codespace/code.go:
```diff
@@ -12,30 +12,40 @@ func newCodeCmd(app *App) *cobra.Command {
 
 	codeCmd.Flags().StringVarP(&codespace, "codespace", "c", "", "Name of the codespace")
 	codeCmd.Flags().BoolVar(&useInsiders, "insiders", false, "Use the insiders version of Visual Studio Code")
+	codeCmd.Flags().BoolVar(&useWeb, "web", false, "Use the web version of Visual Studio Code")
```
Should this be a `BoolVarP` instead so we can accept a shorthand `-w`?

That would be consistent with the `pr view -w` and `repo view -w` options that are already supported (and probably others as well).
```suggestion
	codeCmd.Flags().BoolVarP(&useWeb, "web", "w", false, "Use the web version of Visual Studio Code")
```

### References
https://github.com/cli/cli/blob/f7ef503051ff7a3845738c4846d73a49f3c62f76/pkg/cmd/pr/view/view.go#L71
https://github.com/cli/cli/blob/f7ef503051ff7a3845738c4846d73a49f3c62f76/pkg/cmd/repo/view/view.go#L70

luanzeba commented on pkg/cmd/codespace/code.go:
```diff
@@ -12,30 +12,40 @@ func newCodeCmd(app *App) *cobra.Command {
 		return err
 	}
 
+	if useWeb {
+		if err := a.browser.Browse(codespace.WebUrl); err != nil {
+			return fmt.Errorf("error opening Codespace: %w", err)
+		}
```
Both in `pr view -w` and `repo view -w`, we post a message to the user if we detect a TTY. It seems like it would be nice to be consistent here.
```go
if useWeb {
	openURL := codespace.WebUrl
	if a.io.IsStdoutTTY() {
		fmt.Fprintf(a.io.ErrOut, "Opening %s in your browser.\n", utils.DisplayURL(openURL))
	}
	if err := a.browser.Browse(openURL); err != nil {
		return fmt.Errorf("error opening Codespace: %w", err)
	}
}
```

Although we'd need to import `"github.com/cli/cli/v2/utils"` at the top to accomplish this

### References
https://github.com/cli/cli/blob/f7ef503051ff7a3845738c4846d73a49f3c62f76/pkg/cmd/pr/view/view.go#L103-L109
https://github.com/cli/cli/blob/f7ef503051ff7a3845738c4846d73a49f3c62f76/pkg/cmd/repo/view/view.go#L136-L141

luanzeba reviewed (commented):
Nice work!! I added a couple of questions/suggestions based on other implementations of opening a resource on the browser when using the gh cli

josebalius commented on pkg/cmd/codespace/code.go:
```diff
@@ -12,30 +12,40 @@ func newCodeCmd(app *App) *cobra.Command {
 		Args:  noArgsConstraint,
 		RunE: func(cmd *cobra.Command, args []string) error {
-			return app.VSCode(cmd.Context(), codespace, useInsiders)
+			return app.VSCode(cmd.Context(), codespace, useInsiders, useWeb)
```
This seems fine to me. I wonder how much usage this will get and whether `gh codespace web` might be warranted but we can always introduce that later.

josebalius reviewed (commented):
Nice job! @luanzeba had some good feedback so I'll defer to their approval.

jackcmeyer committed:
support -w shorthand flag

Co-authored-by: Luan Vieira <luanzeba@github.com>

jackcmeyer committed:
print opening in browser message when in tty

jackcmeyer commented:
> I added a couple of questions/suggestions based on other implementations of opening a resource on the browser when using the gh cli

Thanks for the review, @luanzeba. I've implemented the suggestions. I appreciate you catching the inconsistencies with other features.

luanzeba commented on pkg/cmd/codespace/code.go:
```diff
@@ -5,37 +5,50 @@ import (
 		Args:  noArgsConstraint,
 		RunE: func(cmd *cobra.Command, args []string) error {
-			return app.VSCode(cmd.Context(), codespace, useInsiders)
+			return app.VSCode(cmd.Context(), codespace, useInsiders, useWeb)
```
I just realized that we are not respecting the `useInsiders` flag in case someone provides both that + the new `web` option. TIL that [VS Code for Web has an insiders build](https://code.visualstudio.com/docs/editor/vscode-web#_relationship-to-vs-code-desktop) at https://insiders.vscode.dev/

Although it doesn't look like that works for Codespaces when I tried it out. We should probably add a validation that checks only one of those flags is present otherwise provide an error to the user right above this.

cmbrose replied:
There is an optional query param `vscodeChannel` that you can use to select the between `insiders` and `stable` - e.g. `https://<name>.github.dev/?vscodeChannel=insiders` - in order to handle the `--insiders` flag. I think you'd be able to  do something like `codespace.WebUrl + "?vscodeChannel=insiders"` below.

luanzeba commented on pkg/cmd/codespace/code.go:
```diff
@@ -5,37 +5,50 @@ import (
+		}
+		if err := a.browser.Browse(openURL); err != nil {
+			return fmt.Errorf("error opening Codespace: %w", err)
+		}
```
I believe we need to return right here otherwise the function will continue executing, correct? (I'm new to go so feel free to correct me if I'm wrong)
I see that other implementations (see below) just return the `Browse` function directly but I like that you chose to handle any errors there. Maybe you can bring back the return nil you had previously

https://github.com/cli/cli/blob/dd1dd05354535ed06e43652badf23714de50a8f7/pkg/cmd/codespace/code.go#L54

https://github.com/cli/cli/blob/f7ef503051ff7a3845738c4846d73a49f3c62f76/pkg/cmd/repo/view/view.go#L140

luanzeba reviewed (commented):
Thanks for the quick turnaround, nice work! Sorry I had one thing I missed in my previous review.

By the way, I don't have approval rights in this repo, but hopefully someone from the appropriate team will review this soon

mislav committed:
Merge remote-tracking branch 'origin' into open-codespace-in-web

mislav committed:
:nail_care:

mislav reviewed (approved):
Great job, everyone. I have pushed tweaks based on the latest review so that this can go out in the next release 🤘
